### PR TITLE
Individual Assignments

### PR DESCRIPTION
**Pull request by tarebyte**
This will allow an educator to:

* Create a new individual assignment
* Give their students an invitation url
* Allow students
  - Access the url
  - Add them to the organization
  - Create their personal team
  - Create the assignment repository

### CODE REVIEW

**Comment by tarebyte** [comment]
I'm going use `Assignment` and `GroupAssignment` if nothing else than just because I have typing individual all of the time.

File: app/models/assignment_invitation.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+class AssignmentInvitation < ActiveRecord::Base
+  belongs_to :assignment
+
+  validates_presence_of   :key
+  validates_uniqueness_of :key
+
+  after_initialize :assign_key
+
+  def redeem(user)
```
Comment (tarebyte) [inline app/models/assignment_invitation.rb]: This seems like a lot for one method, I'll probably break this up into a few smaller parts. @johndbritton do you think this should stay on the model, or should these become service objects?

File: app/models/assignment_invitation.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    repo_access = user.repo_accesses.find_or_create_by(organization: organization)
+
+    if repo_access.new_record?
+      repo_access.create_github_team(assignment_owner, "Team: #{Time.now}")
```
Comment (tarebyte) [inline app/models/assignment_invitation.rb]: I'm also open to suggestions about naming assignments and repos

File: app/models/assignment_invitation.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    end
+  end
+
+  def find_or_create_assignment_repo_by_repo_access(repo_access)
+    assignment_repo = assignment.assignment_repos.find_or_create_by(repo_access: repo_access)
+
+    if assignment_repo.new_record?
+      assignment_repo.repo_access = repo_access
+
+      new_repo_name = "#{assignment.title} #{Time.zone.now}"
```
Comment (tarebyte) [inline app/models/assignment_invitation.rb]: @johndbritton any feedback for a better name would be much appreciated!

File: app/models/assignment_invitation.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    assignment_repo
+  end
+
+  def find_or_create_repo_access_by_organization(user, organization)
+    repo_access = user.repo_accesses.find_or_create_by(organization: organization)
+
+    if repo_access.new_record?
+      repo_access.create_github_team(assignment_owner, "Team: #{Time.zone.now}")
```
Comment (tarebyte) [inline app/models/assignment_invitation.rb]: Also naming help here

File: app/controllers/assignment_invitations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class AssignmentInvitationsController < ApplicationController
+  before_action :authenticate_with_pre_login_destination, only: [:show]
+
+  def show
+    @invitation = AssignmentInvitation.find_by_key!(params[:id])
+
+    if @invitation.redeem(current_user)
+      render :success, layout: false, status: 200
```
Comment (tarebyte) [inline app/controllers/assignment_invitations_controller.rb]: right now I'm just returning a layout with `success` for `failed`, obviously these need to be more descriptive. For success what do you think about maybe giving the repo url. For failure, probably just better copy.

**Comment by tarebyte** [comment]
@johndbritton this is how adding assignment looks and works right now. I'm definitely up for feedback before I finished writing the tests.

assignment_invitations

File: app/models/assignment_invitation.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
+class AssignmentInvitation < ActiveRecord::Base
+  belongs_to :assignment
+
+  validates :key, presence:   true
+  validates :key, uniqueness: true
+
+  after_initialize :assign_key
+
+  def redeem(user)
+    repo_access     = find_or_create_repo_access_by_organization(user, assignment.organization)
```
Comment (tarebyte) [inline app/models/assignment_invitation.rb]: @johndbritton I'd like to also get your opinion on these two methods, not sure if I should leave them on the model or extract them out to another service object. Feeling like I should leave them here

**Comment by tarebyte** [comment]
@johndbritton this works for the current form of organizations, however the API endpoints are broken for the new beta orgs

File: app/models/git_hub_repository.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    if (repo = org_owner.github_client.create_repository(repo_name, options))
+      GitHubRepository.new(repo.id)
+    else
+      NullGitHubRepository.new
+    end
+  end
+
+  def self.github_repo_options(github_organization, team_id)
+    {
```
Comment (tarebyte) [inline app/models/git_hub_repository.rb]: not sure how I feel about this, maybe this should be part of a form that the user gets when creating the assignment?

Comment (tarebyte) [inline app/models/git_hub_repository.rb]: @johndbritton how do feel about these being attributes on the `Assignment` model?

**Comment by tarebyte** [comment]
@bkeepers I've signed the CLA, is this PR failing that check because it was open before you added it?

**Comment by bkeepers** [comment]
@tarebyte It should update all of your open pulls as soon as you accepted it, so I'm not sure what happened. I gave it a swift kick though and it should be good now.

**Comment by tarebyte** [comment]
:heart_eyes_cat:

**Comment by tarebyte** [comment]
It lives :tada: 

I still need to update the template so that they students doesn't see the normal classroom template.

And I need to update the copy.

assignment_invitations_final

File: spec/models/null_github_team_spec.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+describe NullGitHubTeam do
+  it 'exposes the same public interface as GitHubTeam' do
+    expect(described_class).to match_the_interface_of(GitHubTeam)
```
Comment (johndbritton) [inline spec/models/null_github_team_spec.rb]: TIL about `match_the_interface_of`

File: spec/models/null_git_hub_repository_spec.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+describe NullGitHubRepository do
+  it 'exposes the same public interface as GitHubTeam' do
+    expect(described_class).to match_the_interface_of(NullGitHubRepository)
```
Comment (johndbritton) [inline spec/models/null_git_hub_repository_spec.rb]: Did you mean to match `GitHubRepository` here?

File: spec/models/null_git_hub_repository_spec.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+describe NullGitHubRepository do
+  it 'exposes the same public interface as GitHubTeam' do
```
Comment (johndbritton) [inline spec/models/null_git_hub_repository_spec.rb]: copy pasta mistake?

File: spec/models/null_github_team_spec.rb (hunk 8 above)
Comment (johndbritton) [inline spec/models/null_github_team_spec.rb]: Actually, looks like a custom matcher per https://robots.thoughtbot.com/testing-null-objects

**Comment by johndbritton** [comment]
:zap: Nice work

File: spec/models/null_git_hub_repository_spec.rb (hunk 10 above)
Comment (tarebyte) [inline spec/models/null_git_hub_repository_spec.rb]: yup!

File: spec/models/null_github_team_spec.rb (hunk 8 above)
Comment (tarebyte) [inline spec/models/null_github_team_spec.rb]: That's exactly the article I used to create that

File: spec/models/null_git_hub_repository_spec.rb (hunk 9 above)
Comment (tarebyte) [inline spec/models/null_git_hub_repository_spec.rb]: Indeed

File: spec/models/null_github_team_spec.rb (hunk 8 above)
Comment (johndbritton) [inline spec/models/null_github_team_spec.rb]: The Googles are strong with this one.

**Pull request merged**